java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodePropertiesProvider.java

```diff
@@ -0,0 +1,210 @@
+        return Collections.emptyList();
+    }
+
+    @Override
+    public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
+        if (!COMMANDS.contains(command)) {
+            return CompletableFuture.completedFuture(null);
+        }
+        if (arguments == null || arguments.isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
+        TreeNodeRegistry r = Lookup.getDefault().lookup(TreeNodeRegistry.class);
+        if (r == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+        int nodeId = ((JsonPrimitive) arguments.get(0)).getAsInt();
+        TreeViewProvider nodeProvider = r.providerOf(nodeId);
+        Node node = null;
+        if (nodeProvider != null) {
+            node = nodeProvider.findNode(nodeId);
+        }
+        if (node == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+        boolean getProps = COMMAND_GET_NODE_PROPERTIES.equals(command);
+        Node.PropertySet[] propertySets = node.getPropertySets();
+
+        if (getProps) {
+            return CompletableFuture.completedFuture(getAllPropertiesMap(propertySets));
+        }
+        if (arguments.size() == 2) {
+            Object propJson = arguments.get(1);
+            if (propJson instanceof JsonNull) {
+                return CompletableFuture.completedFuture(null);
+            }
+            List m = gson.fromJson((JsonElement) propJson, List.class);
+            setAllProperties(propertySets, m);
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    Map<String, ?>[] getAllPropertiesMap(Node.PropertySet[] propertySets) {
+        Map<String, Object>[] allPropertiesMap = new Map[propertySets.length];
+
+        for (int i = 0; i < propertySets.length; i++) {
+            Map<String, Object> propertiesMap = new HashMap<>();
+            Node.PropertySet ps = propertySets[i];
+            propertiesMap.put(PROP_NAME, ps.getName());
+            propertiesMap.put(PROP_DNAME, ps.getDisplayName());
+            propertiesMap.put(PROP_HTML_NAME, ps.getHtmlDisplayName());
+            propertiesMap.put(PROP_SHORT_NAME, ps.getShortDescription());
+            propertiesMap.put(PROP_PREF, ps.isPreferred());
+            propertiesMap.put(PROP_EXPERT, ps.isExpert());
+            propertiesMap.put(PROP_HIDDEN, ps.isHidden());
+            propertiesMap.put(PROPS, getProperties(ps.getProperties()));
+            allPropertiesMap[i] = propertiesMap;
+        }
+        return allPropertiesMap;
+    }
+
+    private Map<String, Object>[] getProperties(Node.Property<?>[] properties) {
+        Map<String, Object>[] props = new Map[properties.length];
+        for (int i = 0; i < properties.length; i++) {
+            Map<String, Object> propMap = new HashMap<>();
+            Node.Property<?> property = properties[i];
+            if (property.canRead()) {
```

Comment:
This will make some entries of the returned array `null`, client having to check for null/undefined there. Wouldn't it be better to accumulate in a List and turn to array at the end ?